File: lists/3.go
package main

import "fmt"

// Node comment
type Node struct {
	value int
	next  *Node
}

var root = new(Node)

func addNode(node *Node, value int) int {

	if root == nil {
		node = &Node{value, nil}
		root = node
		return 1
	}

	if node.next == nil {
		node.next = &Node{value, nil}
		return 1
	}

	for node.next != nil {
		node = node.next
	}

	node.next = &Node{value, nil}
	return 1
}

func printList(node *Node) {

	if node == nil {
		fmt.Println(" Empty List")
		return
	}

	for node != nil {
		fmt.Printf("%d -->", node.value)
		node = node.next
	}
	fmt.Println()
}

func reverseLL(node *Node) int {

	if node == nil {
		fmt.Println(" Empty List")
		return -1
	}

	var prev *Node = nil
	var temp *Node = nil
	var current = node

	for current != nil {
		temp = current.next
		current.next = prev
		prev = current
		current = temp
	}
	node = prev

	fmt.Print("Reversed LL ---> ")
	for node != nil {
		fmt.Printf("%d -->", node.value)
		node = node.next
	}
	fmt.Println()
	return 1
}

func main() {
	root = nil
	addNode(root, 11)
	addNode(root, 22)
	addNode(root, 33)
	addNode(root, 44)
	addNode(root, 55)
	fmt.Print("Original LL ---> ")
	printList(root)
	reverseLL(root)
}

File: lists/3.js
const head = Symbol("head");

class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this[head] = null;
  }
  add(data) {
    const newNode = new Node(data);
    if (this[head] === null) {
      this[head] = newNode;
    } else {
      let current = this[head];
      while (current.next != null) {
        current = current.next;
      }
      current.next = newNode;
    }
  }
  print() {
    if (this[head] === null) {
      console.log("Empty List");
      return
    }
    let current = this[head];
      while (current != null) {
        console.log(current.data, "->");
        current = current.next;
      }

  }
  reverseList() {
    if (this[head] === null) {
      console.log("Empty List");
      return
    }
    let prev = null;
    let temp = null;
    let current = this[head];

    while (current != null) {
      temp = current.next;
      current.next = prev;
      prev = current;
      current = temp;
    }

    let newHead = prev;
    while (newHead != null) {
      console.log(newHead.data, "->");
      newHead = newHead.next;
    }

  }

}

const list = new LinkedList();

list.add(1);
list.add(2);
list.add(3);
list.add(4);
list.print();
list.reverseList();
